### DISCUSSION TITLE
Expanding exoplanet features (3D exoplanets; zooming into exoplanets; exoplanet skies, surfaces and orbits; etc.)

### DISCUSSION THREAD

**Discussion by ghost** [Stellarium/stellarium#3581] [Ideas]
Hello there. :-)

I'm new to this place and to Stellarium.

I'd like to make a few suggestions that I believe would greatly improve Stellarium and set it apart from other planetarium/astronomy software. Perhaps I could help in adding these features myself, although I should say I have no past experience in coding. >_<

So I was thinking, what if Stellarium could allow us to "zoom into" exoplanetary systems? I realize there's no telescope that can do that today---at least not until quantum-assisted telescopes become a thing in the far future---but I think it would be a great educational and fun feature to have. If I recall correctly, someone tried adding a feature like that on his own lesser-known app several years ago, but I couldn't remember which app it was. I believe his app is defunct now (i.e. no longer in development). So I was thinking, maybe Stellarium should have this feature, considering this app is popular, and its development is still taking place.

So maybe we could zoom into the exoplanetary systems until we could see the exoplanets in question, in addition to their hypothetical surface features. And then maybe we could also see these exoplanets orbit around their host stars as time goes by, etc. Also, maybe we could also see what the night sky looks like on the surfaces of these exoplanets.

I tried searching if someone already did this on Stellarium, and I did come across a few YouTube videos from many years back that show the orreries of several exoplanetary systems:

https://www.youtube.com/watch?v=YMLus46vGHM

I realize apps like SpaceEngine and NASA's Eyes already exist, but I'd love it if Stellarium expands its exoplanet features as well. The more choices we have, the better the products/services would become. The more the merrier, as they say. :-)

Maybe these features could become part of a paid subscription or something to that effect, in order to make the program financially sustainable. Or perhaps Stellarium Labs could add these extra exoplanet features for the paid version of their mobile and VR apps. I noticed on the Stellarium Labs homepage that a VR version is being prepared for Meta's headsets, and it seems from the screenshots that 3D stars will be featured for the first time. So maybe we could also include 3D exoplanets with the paid versions instead of the free version?

I'm ready to pay for such extra features, and also willing to help with the coding. In fact, I really wanna learn how to add extra features to Stellarium. 

Before I shut up, I found an interesting software that simulates what exoplanetary systems would look like with future space telescopes that use starshades:

http://sister.caltech.edu/

Interested to know what you guys think. :-)

--Nadia

**Comment by gzotti** [comment]
I am far less enthusiastic about this. For most EPs (with light curves only) this will remain fantasy until any information can be worked out about the size and orientation of orbits. The angular sizes of Exo systems will likely be too small to be presentable in the context of this program without running into numerical problems. Scaling up by 1000x or more may be required, which may then look awkward with other stars nearby. 

Even worse, hypothetical surface features. These are "artist concepts" whenever displayed, and would promise false claims. Yes, SpaceEngine is the program that comes into my mind for approaching/visualizing these objects. 

Despite my objections, I won't stop you adding them in the EP plugin, as long as these features are optional. 

IMHO simulating the orbits of close binaries would be far more important.

**Reply by ghost** [reply]
Hello Mr. Georg,

Thanks for the reply. <3

I must confess that my original intention was to merely suggest these extra exoplanet features to the developers of Stellarium and to offer financial support for the inclusion of such features, considering that my knowledge in coding is literally zero lol. >_<

I'm under the impression that there are 2 different Stellarium projects. Am I correct to assume that Stellarium Labs have different visions/goals than the free open-source version of Stellarium? For example, Stellarium Labs's products (Stellarium Mobile Plus and, presumably, Stellarium VR) seem to offer interactive 3D views of the Solar System, a paid mobile feature that isn't available on the desktop version of Stellarium. Moreover, their products come in 2 versions: a free version and a paid version.

A day before creating my Github account, I sent an email to Stellarium Labs in which I expressed my full commitment/willingness to finance and pay for a 3D exoplanets feature for the paid version of their Mobile Plus and VR apps. 

At the same time, however, I also thought about asking for the inclusion of these features in the open-source desktop version of Stellarium as well, which is why I made my Github account.

I hope someone in the future adopts my idea and adds such features into Stellarium, and I'd be more than willing to help finance whoever is open to the idea. Perhaps this would pave the way for a new economic model for the open-source version of Stellarium, in which people with ideas for the program and no knowledge of coding could finance their ideas, which would help make Stellarium financially self-sustainable.

That said, despite my complete lack of knowledge in coding, I feel very motivated to try adding such features myself. The only problem is I have no idea where to even start. I looked at the codes of the Exoplanets plugin today on notepad. I guess I'll start playing around with the code and see what it does. Hopefully my PC doesn't crash lol.

I have a number of arxiv.org and IOPScience papers to make sure the exoplanets' orbits and surfaces are represented as scientifically accurate as possible. My bookmarks are filled with thousands of such papers.

With regard to the scaling issue, this is where my lack of coding knowledge becomes very apparent. Is it possible to blacken out all the nearby stars when zooming into an exoplanetary system? For example, the more you zoom into the star that hosts the exoplanets (e.g. Trappist-1), the more the nearby stars "fade away". This way, it won't look awkward, as you implied.

Or maybe, instead of zooming into the systems, there's a way to click on the star and get a pop-up that shows an interactive 3D orrery of the exoplanetary system, as well as 3D models of the exoplanets themselves? This would eliminate the awkwardness of having infinite zooms, and would also eliminate the problem of the system's inclination relative to the earth.

Sorry for the long post.

--Nadia

**Reply by gzotti** [reply]
Yes, there are two separate projects called Stellarium, both founded by Fabien. The open-source version discussed here is maintained and further developed by (currently) three people, after a long time of being usually just two. We are (I think) mostly occupied by our own projects and ideas. Thank you for your offer of financing such development, but this would still need a contract with some host institute. (PM for details.) 

I still don't see displaying EPs in the wider context of neighbouring stars, constellations or the night sky which is the focus of Stellarium. Your description looks like an extra that could interact with a separate web browser window (even on a separate screen) which might have its own 3D viewer with some orbiting colored spheres which you could turn around and view from all sides. Maybe show a button "show from earth" which re-aligns the view with geocentric equatorial coordinates. Next to that would be a textual description of what the user can see here and what makes that respective system so interesting. 

The user handling could go from the current EP plugin which shows symbols at the locations in the sky where EPs have been detected. Select one EP, then press some hotkey to load further info in that external browser. 

Now, if such website already existed, with a public API that can be called by an URL that includes some object ID, this could be an easy mission. The biggest part would be setting up such website. It would be mostly independent from Stellarium, and everybody could find EP info there. Whoever cares for the website can maintain/update data, and we Stellarium developers would just extend the EP plugin with such said call of that website URL. What then would still be required would be some way to download a file which indicates which EPs are handled by that website. These could then be shown in an different color to say "more data online". 

EDIT: Or, after all, we can just use/extend the OnlineQueries plugin for that purpose, looking up whatever info is provided for a selected EP system. Maybe connecting that somehow to the EP plugin if at all necessary. 

The same web service, if allowed, could probably be also embedded in, or interact with, the web version or other astro software.

**Reply by ghost** [reply]
Thanks for the reply, Mr. Zotti.

This seems to be very complicated for a beginner such as myself. 

May I ask a beginner's question? I've been looking at the files and folders of Stellarium and came across "ssystem_major.ini". Am I correct in assuming that this is the file that is responsible for making all the planets and their orbits/movements/locations appear on Stellarium?

If so, then is it programmably possible for me to create a similar file in the Exoplanets plugin folder for the orbits of the exoplanets? 

I've been looking at the zoom/scaling issue lately, and I actually think this could be pulled off in a way that won't make it look awkward. Since I'm a beginner, I might have to give up on the hypothetical surfaces feature and just focus on making exoplanets look like faint dots of light next to their host stars. And perhaps these faint dots could be seen orbiting around their stars as time goes by.

I might do it for planets that have already been directly imaged via high contrast imaging. I'd like to see if I could recreate something like the below gif on stellarium:

https://en.wikipedia.org/wiki/File:HR_8799_Orbiting_Exoplanets.gif

I already located the star in question on Stellarium (HR 8799). As I zoom into the star, perhaps I could make its brightness fade away a little bit, revealing tiny faint dots of the exoplanets that are orbiting it.

Thanks for answering my previous questions/points. I had no idea desktop Stellarium had such few active developers. Count me as a 4th active developer once I learn how to code lol. >_<

**Reply by gzotti** [reply]
Yes, this file contains data for the Solar planets which is the central interest of most users of this program. It is processed by a module dedicated to simulation of _**our**_ Solar System. Seriously, EPs are currently a side note like quasars or pulsars. The only notable comment I can remember so far was a complaint by somebody who thought that massive cluster of EPs [indicating those discovered by Kepler] must be a bug, and he "cannot use such meaningless visualisation of the EP distribution [with such a stupid blob in the sky]" (or similar). 

Defining a file for EP orbits will not help until a dedicated module will have been developed which would read and process such file. To get an idea of the program, you should read the User Guide which describes the structure of user-accessible data files, and you must install a build environment as described on these pages and probably compile&run the program, and make some experimental changes, to fully understand the internals. 

Seriously, it may be simpler to go the web way as described, find a simple Keplerian solar system simulator engine (my favourite search engine finds quite a handful... There are even already EP engines and whole EP websites available! - just extend with whatever you need and add an API to call the site by object name!) and parametrize this with EP data. Javascript is also regarded simpler than C++ by most people. Once such website exists, optimally with a menu of its own to select objects, and esp. detailed information about the objects (can also include actual photos, links to papers, ...), it could probably be accessed by Stellarium (or other software) to provide add-on information.

**Reply by ghost** [reply]
>Yes, this file contains data for the Solar planets which is the central interest of most users of this program. It is processed by a module dedicated to simulation of our Solar System. Seriously, EPs are currently a side note like quasars or pulsars. The only notable comment I can remember so far was a complaint by somebody who thought that massive cluster of EPs [indicating those discovered by Kepler] must be a bug, and he "cannot use such meaningless visualisation of the EP distribution [with such a stupid blob in the sky]" (or similar).

I understand. I realize that for many people, simulating exoplanets isn't important. However, as long as such features are optional, then having them could expand Stellarium's outreach. I mean, at the end of the day, this is a planetarium/astronomy software after all, and simulating the night sky could include a plethora of different features, including non-earthly features.

I'm probably getting ahead of myself right now, but last night I thought that maybe one day, once I can improve my coding skills, I could add features to Stellarium that simulate what the night sky would look like with various current and upcoming space telescopes, including the Habitable Worlds Observatory, etc. So I have many of these ideas that I think are not only fun and educational, but would also make Stellarium go leaps and bounds beyond any other competing software/app.

So why not? I mean, as long as you guys are okay with it and as long as these are just optional features, then I think it'll be nice to dream. And if I fail to code these features, I'm willing to go to great lengths to fund the developers and perhaps even find a sponsor for Stellarium. Maybe a world-renowned space body or university, etc.

I hope I'm not coming across as pushy. I totally understand that this isn't my software and that the decision is not mine. But I'm grateful that you've been open-minded about my ideas. I reaffirm that I in no way wish to impose anything that undermines what Stellarium is primarily about, and that everything I've suggested so far should be optional for the end user.

>Defining a file for EP orbits will not help until a dedicated module will have been developed which would read and process such file. To get an idea of the program, you should read the User Guide which describes the structure of user-accessible data files, and you must install a build environment as described on these pages and probably compile&run the program, and make some experimental changes, to fully understand the internals.

Ah, I see. Thanks for the info. I had no idea I needed to create a module to handle this task. Could you please tell me the name of the file/module that runs the movements of the solar system planets? I'd like to take a look at the code and see what I could learn from it. Maybe I could teach myself how to make a new module and add it to Stellarium etc.

----------------------------------

EDIT: Maybe I'll just make the faint exoplanets static until I figure out how to create a module that performs their movements/orbits.

**Reply by gzotti** [reply]
There are many ideas what else can be added to the project. Usually after accepting some contribution, it means the developer leaves and maintaining it must be made by other people. This means, the maintainers must at least have a rough understanding of the ever-increasing code, which makes me reluctant about adding too much unknown stuff for too few maintainers and maybe too few users. But this should not stop anybody from developing something solid, without us having to do the work instead of our own aims. I am also not the only one to decide about acceptance.

You can already show non-optical surveys in HiPS format. Given an ever increasing number of available HiPS, It would be welcome to define/recommend a "shortlist" of the most interesting HiPS per wavelength or sorted by other means. 

The module in charge of computing the Solar system is aptly called SolarSystem. It interacts with objects of the Planet class and its subclasses, MinorPlanet and Comet. All objects move on Orbits, be that their own specialized solutions or simple KeplerOrbits. This is very much the requirement of our Solar System, with light time correction and other refinements I don't see necessary for EPs, so the ExoPlanetarySystem class will likely have to be designed from scratch, with only little code re-use. (Maybe the KeplerOrbit class can be reused, though, or a simple derivative of it.) Effectively, what has to be designed could likewise be used for binaries in orbit, which I would find more important for now on my personal list of things to be still done. (No, I will not start with that in the next weeks!)

**Comment by alex-w** [comment]
My 2 cents:
1) In your youtube video author replaced Solar system data by TRAPPIST-1 data - details is [here](https://astropolis.pl/topic/57770-system-planetarny-trappist-1-w-programie-stellarium/) - it's a very specific simulation with fake magnitudes and textures.
2) Exoplanetary systems can be easy simulated by Celestia project.

To true simulating the exoplanetary systems in Stellarium we should deep refactor at least support star catalogs (need a true 3D places of stars) and movements (flying to the stars - hello quaternions).

Of course we can expand the OnlineQuery plugin by one or few specific websites for exoplanets to propose users viewing details for exoplanets (include orbits, if necessary).
